diff --git a/tools/hls-verifier/include/HlsVhdlTb.h b/tools/hls-verifier/include/HlsVhdlTb.h
index 318b0c4ad..bf1038962 100644
--- a/tools/hls-verifier/include/HlsVhdlTb.h
+++ b/tools/hls-verifier/include/HlsVhdlTb.h
@@ -171,8 +171,7 @@ end
 // Equivalent of VHDL gen_sim_latency_proc
 always @(posedge tb_clk) begin
     if (tb_global_valid && tb_global_ready) begin
-        $display("Simulation done! Latency = %0d cycles",
-                 ($time - RESET_LATENCY) / (2 * HALF_CLK_PERIOD));
+      $display("Simulation done! Latency = %0d cycles", $floor(($time - RESET_LATENCY) / (2 * HALF_CLK_PERIOD)));
     end
 end
   
diff --git a/tools/hls-verifier/lib/HlsVhdlTb.cpp b/tools/hls-verifier/lib/HlsVhdlTb.cpp
index 678223dce..205566abb 100644
--- a/tools/hls-verifier/lib/HlsVhdlTb.cpp
+++ b/tools/hls-verifier/lib/HlsVhdlTb.cpp
@@ -446,7 +446,8 @@ void getConstantDeclaration(mlir::raw_indented_ostream &os,
     c.declareConstants(os, ctx, inputVectorPath, outputFilePath);
   }
   declareConstant(ctx, os, "HALF_CLK_PERIOD", TIME, "2.00");
-  declareConstant(ctx, os, "RESET_LATENCY", TIME, "10.00");
+  declareConstant(ctx, os, "RESET_LATENCY", TIME, "8.00");
+
   declareConstant(ctx, os, "TRANSACTION_NUM", INTEGER, to_string(1));
 }
 
diff --git a/tools/hls-verifier/resources/templates_verilog/template_single_argument.v b/tools/hls-verifier/resources/templates_verilog/template_single_argument.v
index 88d172e5e..e9314df18 100644
--- a/tools/hls-verifier/resources/templates_verilog/template_single_argument.v
+++ b/tools/hls-verifier/resources/templates_verilog/template_single_argument.v
@@ -85,8 +85,9 @@ initial begin : file_to_mem
             read_token(fp, token);
 
             // wait for done
-            @(posedge clk);
-            wait (done);
+            while (done != 1'b1) begin
+              @(posedge clk);
+            end
 
             // read data
             read_token(fp, token);
@@ -177,4 +178,4 @@ initial begin : mem_to_file
     end
 end
 
-endmodule
\ No newline at end of file
+endmodule
